### PR TITLE
add CPU player and turn function

### PR DESCRIPTION
**Pull request by AndySch666**
- add a CPU player
- add a turn function

currently the game have no ending, sorry about that i will fix with the new posted user story.

### CODE REVIEW

**Comment by ThomasKirst** [comment]
Hi Andi,
I get this warning when starting the app:
```
Warning: Invalid DOM property `crossorigin`. Did you mean `crossOrigin`?
    at link
    at head
    at Head (webpack-internal:///./node_modules/next/dist/pages/_document.js:349:1)
```

When I change `crossorigin` to `crossOrigin` I get:
```
Warning: Received `true` for a non-boolean attribute `crossOrigin`.

If you want to write it to the DOM, pass a string instead: crossOrigin="true" or crossOrigin={value.toString()}.
```
Try to fix this ;)

**Comment by ThomasKirst** [comment]
Please also check for your `format on save` setting in vs-code.

File: components/XOGame.js [hunk 1]
```diff
@@ -1,14 +1,25 @@
 import styled from "styled-components";
-import {useState} from "react";
+import {useEffect, useState} from "react";
 
 const players = {
-
+  CPU: {
+    SYM: "O",
```
Comment (ThomasKirst) [inline components/XOGame.js:6]:
```suggestion
    SYMBOL: "O",
```
Why save three bytes? Better readability for the win!

File: components/XOGame.js [hunk 2]
```diff
@@ -17,22 +28,54 @@ export default function XOGame() {
...
     board[arrayIndex][index] = players?.HUMAN?.SYM;
     setBoard((board) => [...board]);
     
-    
+    setIsCPUNext(true);
+  }
+
+  useEffect(() => {
+    if (isCPUNext) {
+      cPUPlay();
```
Comment (ThomasKirst) [inline components/XOGame.js:44]:
```suggestion
      cpuPlay();
```

File: components/XOGame.js [hunk 3]
```diff
@@ -17,22 +28,54 @@ export default function XOGame() {
...
+
+  function cPUPlay() {
+    sleep(1000);
+
+    const cPUMove = getCPUTurn();
+
+    board[cPUMove.arrayIndex][cPUMove.index] = players?.CPU?.SYM;
+
+    setBoard((board) => [...board]);
+    setIsCPUNext(false);
```
Comment (ThomasKirst) [inline components/XOGame.js:56]:
```suggestion
  function cpuPlay() {
    sleep(1000);

    const cpuMove = getCPUTurn();

    board[cpuMove.arrayIndex][cpuMove.index] = players?.CPU?.SYM;

    setBoard((board) => [...board]);
    setIsCPUNext(false);
```

File: components/XOGame.js [hunk 4]
```diff
@@ -17,22 +28,54 @@ export default function XOGame() {
...
   }
 
+  function getCPUTurn() {
+    const emptyIndexes = [];
+    board.forEach((row, arrayIndex) => {
+      row.forEach((cell, index) => {
+        if (cell === "") {
+         emptyIndexes.push({arrayIndex, index});
+        }
+      });
```
Comment (ThomasKirst) [inline components/XOGame.js:66]:
```suggestion
    board.forEach((row, rowIndex) => {
      row.forEach((cell, columnIndex) => {
        if (cell === "") {
         emptyIndexes.push({ rowIndex, columnIndex });
        }
      });
```

File: components/XOGame.js [hunk 5]
```diff
@@ -17,22 +28,54 @@ export default function XOGame() {
...
+  }
+
+  useEffect(() => {
+    if (isCPUNext) {
+      cPUPlay();
+    }
+  }, [isCPUNext]);
+
+  function cPUPlay() {
+    sleep(1000);
```
Comment (ThomasKirst) [inline components/XOGame.js:49]: It's generally not advisable to block the JS execution.
I would rather use a `setTimeout` function within the `useEffect()`, something like:
```
useEffect(() => {
    if (isCPUNext) {
      setTimeout(cPUPlay, 1000);
    }
  }, [isCPUNext]);
  ```
Then you could remove the `sleep(1000)` part in your `cpuPlay()` function.

As a consequence you might want to check in the player´s turn function that it´s not the CPU´s turn when the player attempts to click a second time within the time window.
```
  function playFieldLocation(arrayIndex, index) {
    if (!isCPUNext) {
```

**Review by ThomasKirst — CHANGES_REQUESTED** [review]
Looks like a promising continuation of your game.
I made a few remarks you might want to consider implementing, especially regarding your app´s performance.
Keep it up 🚀

File: components/XOGame.js (hunk 5 above)
Comment (AndySch666) [reply components/XOGame.js:49]: Thanks for the input

**Review by ThomasKirst — CHANGES_REQUESTED** [review]
Only two small remarks left.

**Review by ThomasKirst — APPROVED** [review]
Ready to merge!

**Pull request merged**